aten/src/ATen/native/mkldnn/xpu/detail/MatmulHelpers.h

```diff
@@ -0,0 +1,231 @@
+      int64_t m,
+      int64_t k,
+      int64_t n,
+      int64_t mb,
+      int64_t bsA,
+      int64_t bsB,
+      bool m2_trans,
+      bool with_bias)
+      : ndim(dim),
+        m(m),
+        k(k),
+        n{n},
+        mb{mb},
+        bsA(bsA),
+        bsB(bsB),
+        m2_trans(m2_trans),
+        with_bias(with_bias) {}
+
+  using arg_map_t = std::unordered_map<int, dnnl::memory>;
+  using md_t = dnnl::memory::desc;
+
+  void handle2D(
+      const at::Tensor& m1,
+      const at::Tensor& m2,
+      const at::Tensor& dst) {
+    m1_dims = {m, k};
+    m2_dims = {k, n};
+    dst_dims = {m, n};
+
+    m1_strides = {m1.stride(0), m1.stride(1)};
+    m2_strides = m2_trans ? std::vector<int64_t>{m2.stride(0), m2.stride(1)}
+                          : std::vector<int64_t>{m2.stride(1), m2.stride(0)};
+    dst_strides = {dst.stride(0), dst.stride(1)};
+  }
+
+  void handle3D(
+      const at::Tensor& m1,
+      const at::Tensor& m2,
+      const at::Tensor& dst) {
+    m1_dims = dnnl::memory::dims({bsA, m, k});
+    m2_dims = dnnl::memory::dims({bsB, k, n});
+    dst_dims = dnnl::memory::dims({mb, m, n});
+
+    m1_strides = {m1.stride(0), m1.stride(1), m1.stride(2)};
+    m2_strides = m2_trans
+        ? std::vector<int64_t>{m2.stride(0), m2.stride(1), m2.stride(2)}
+        : std::vector<int64_t>{m2.stride(0), m2.stride(2), m2.stride(1)};
+    dst_strides = {dst.stride(0), dst.stride(1), dst.stride(2)};
+  }
+
+  void initialize_md(
+      const at::Tensor& m1,
+      const at::Tensor& m2,
+      const at::Tensor& dst) {
+    m1_md = create_md(m1, m1_dims, m1_strides);
+    m2_md = create_md(m2, m2_dims, m2_strides);
+    dst_md = create_md(dst, dst_dims, dst_strides);
+  }
+
+  std::tuple<md_t, md_t, md_t, md_t> query_md() {
+    return {m1_md, m2_md, dst_md, bias_md};
+  }
+
+  void initialize(
+      const at::Tensor& m1,
+      const at::Tensor& m2,
+      const at::Tensor& dst,
+      const at::Tensor& bias) {
+    if (ndim == 2) {
+      handle2D(m1, m2, dst);
+    } else if (ndim == 3) {
+      handle3D(m1, m2, dst);
+    } else {
+      TORCH_CHECK(false, "only support 2D or 3D matmul, got ndim:", ndim);
+    }
+    initialize_md(m1, m2, dst);
+
+    if (with_bias) {
+      bias_dims = get_onednn_dims(bias);
+      bias_strides = get_onednn_strides(bias);
+      bias_md = dnnl::memory::desc(
+          bias_dims, get_onednn_dtype_include_double(bias), bias_strides);
+    } else {
+      bias_md = dnnl::memory::desc();
+    }
+  }
+
+  arg_map_t create_memory(
+      const at::Tensor& m1,
+      const at::Tensor& m2,
+      const at::Tensor& dst,
+      const at::Tensor& bias) {
+    std::unordered_map<int, dnnl::memory> args;
+    auto& eng = GpuEngineManager::Instance().get_engine();
+    args.insert({DNNL_ARG_SRC, dnnl::memory(m1_md, eng, m1.data_ptr())});
+    args.insert({DNNL_ARG_WEIGHTS, dnnl::memory(m2_md, eng, m2.data_ptr())});
+    args.insert({DNNL_ARG_DST, dnnl::memory(dst_md, eng, dst.data_ptr())});
```

Comment:
```suggestion
    args.emplace({DNNL_ARG_SRC, dnnl::memory(m1_md, eng, m1.data_ptr())});
    args.emplace({DNNL_ARG_WEIGHTS, dnnl::memory(m2_md, eng, m2.data_ptr())});
    args.emplace({DNNL_ARG_DST, dnnl::memory(dst_md, eng, dst.data_ptr())});
```
Avoid pair constructor overhead.